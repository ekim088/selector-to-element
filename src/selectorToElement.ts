import makeNodeTree from './makeNodeTree';
import parseAttributes from './parseAttributes';
import parseClasses from './parseClasses';
import parseId from './parseId';
import parsePseudoClasses, {
	parsePseudoParam,
	removePseudoClasses
} from './parsePseudoClasses';
import parseTag from './parseTag';

// input element type guard
const inputTagNames = ['INPUT', 'TEXTAREA', 'SELECT', 'OPTION'];
const isHTMLInputElement = (
	el: HTMLElement | HTMLInputElement
): el is HTMLInputElement => inputTagNames.indexOf(el.tagName) > -1;

// list of known pseudo classes targeting inputs
const inputPseudoClasses = [
	'checked',
	'disabled',
	'invalid',
	'optional',
	'required',
	'valid'
];

/**
 * Transforms a single node selector to an element.
 *
 * @param {string} selector A selector containing a single node.
 * @returns {Element} A HTML Element representing the selector.
 */
export const makeNode = (selector: string): HTMLElement => {
	// parse most attributes in selector with pseudo classes removed so
	// pseudo class param contents like :has() are not captured
	const selectorWithoutPseudos = removePseudoClasses(selector);
	const attributes = parseAttributes(selectorWithoutPseudos);
	const id = parseId(selectorWithoutPseudos);
	const classes = parseClasses(selectorWithoutPseudos);
	const pseudoClasses = parsePseudoClasses(selector);
	const pseudoHas = pseudoClasses.find(pseudo => pseudo.indexOf('has') === 0);
	const tagName = parseTag(selectorWithoutPseudos);

	// check if selector contains a known input-based pseudo class
	const hasInputPseudoClass = pseudoClasses.some(
		pseudoClass => inputPseudoClasses.indexOf(pseudoClass) > -1
	);

	// if selector has input pseudo class and element type has not been
	// specified, then create as input element
	const el = document.createElement(
		hasInputPseudoClass &&
			selector.toLowerCase().indexOf(tagName.toLowerCase()) !== 0
			? 'input'
			: tagName
	);

	if (id) el.id = id;
	if (classes.length > 0) el.className = classes.join(' ');
	Object.entries(attributes).forEach(([name, value]: [string, string]) => {
		el.setAttribute(name, value);
	});

	// support :has() pseudo by appending nodes of value to current node
	if (pseudoHas) {
		const hasVal = parsePseudoParam(pseudoHas);

		// eslint-disable-next-line no-use-before-define
		if (hasVal) selectorToElement(hasVal, el);
	}

	// support input-based pseudo-classes
	if (hasInputPseudoClass && isHTMLInputElement(el)) {
		if (pseudoClasses.indexOf('checked') > -1) {
			if (el.tagName === 'INPUT') {
				el.setAttribute('type', 'checkbox');
				el.checked = true;
			} else if (el.tagName === 'OPTION') {
				el.checked = true;
			}
		}

		if (pseudoClasses.indexOf('disabled') > -1) el.disabled = true;
		if (pseudoClasses.indexOf('invalid') > -1)
			el.setCustomValidity('invalid');
		if (pseudoClasses.indexOf('required') > -1) el.required = true;
	}

	return el;
};

/**
 * Transforms a selector into an HTML element.
 *
 * @param {string} selector A selector.
 * @param {Element} [appendTo] A root element to append the DOM branch specified
 * 	by the selector. Defaults to a new `div`.
 * @returns {Element} A HTML Element representing the selector.
 */
const selectorToElement = (
	selector: string,
	appendTo?: HTMLElement
): HTMLElement => {
	let el: HTMLElement = appendTo || document.createElement('div');
	const nodeTree = makeNodeTree(selector);

	// traverse each level of the node tree and set deeper nodes as children of
	// the previous parent node
	for (let i = 0; i < nodeTree.length; i++) {
		let currentNode: HTMLElement | null = null;

		for (let j = 0; j < nodeTree[i].length; j++) {
			currentNode = makeNode(nodeTree[i][j]);
			el.append(currentNode);
		}

		if (currentNode) el = currentNode;
	}

	return el;
};

export default selectorToElement;
