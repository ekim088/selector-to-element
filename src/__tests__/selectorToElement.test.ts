import selectorToElement from '../selectorToElement';

describe('selectorToElement', () => {
	it('should make an HTML element matching the selector', () => {
		const selectors = [
			'a .b .c',
			'div .a[b=c] ul.list',
			'ol.a li.b.c#test div a article#art1.test[data-test="testing"]',
			'div + span',
			'div ~ span',
			'.test a + #myId.class1 ul.test',
			'article#myId a.linkClass[target="_blank"] img',
			'div .a:nth-child(6)',
			'div [a=b] #c:nth-child(5n + 2)',
			'a#t+article.r img.cl2na span[a="b"]~aside>table'
		];
		selectors.forEach(selector => {
			expect(selectorToElement(selector).matches(selector)).toBeTruthy();
		});
	});

	it('should append branch to a specified root', () => {
		const rootEl = document.createElement('span');
		const selector = 'div + span .a[b=c] ul.list';
		selectorToElement(selector, rootEl);
		expect(rootEl.querySelector(selector)).not.toBeNull();
	});

	it('should support the :has pseudo class', () => {
		const selector =
			'div + span .a[b=c] .test:has(div ul.list li[data-item="testing"]) button';
		const el = selectorToElement(selector);
		expect(el.tagName).toEqual('BUTTON');

		const hasEl = el.closest('.test');
		expect(
			hasEl && hasEl.querySelector('div ul.list li[data-item="testing"]')
		).not.toBeNull();
	});
});
