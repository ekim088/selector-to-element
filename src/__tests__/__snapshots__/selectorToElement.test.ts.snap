// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectorToElement should append ".test a + #myId.class1 ul.test" to the root element 1`] = `
<div
  id="root"
>
  <div
    class="test"
  >
    <a />
    <div
      class="class1"
      id="myId"
    >
      <ul
        class="test"
      />
    </div>
  </div>
</div>
`;

exports[`selectorToElement should append "a .b .c" to the root element 1`] = `
<div
  id="root"
>
  <a>
    <div
      class="b"
    >
      <div
        class="c"
      />
    </div>
  </a>
</div>
`;

exports[`selectorToElement should append "a#t+article.r img.cl2na span[a="b"]~aside>table" to the root element 1`] = `
<div
  id="root"
>
  <a
    id="t"
  />
  <article
    class="r"
  >
    <img
      class="cl2na"
    >
      <span
        a="b"
      />
      <aside>
        <table />
      </aside>
    </img>
  </article>
</div>
`;

exports[`selectorToElement should append "article#myId a.linkClass[target="_blank"] img" to the root element 1`] = `
<div
  id="root"
>
  <article
    id="myId"
  >
    <a
      class="linkClass"
      target="_blank"
    >
      <img />
    </a>
  </article>
</div>
`;

exports[`selectorToElement should append "div + span" to the root element 1`] = `
<div
  id="root"
>
  <div />
  <span />
</div>
`;

exports[`selectorToElement should append "div .a:first-child .b" to the root element 1`] = `
<div
  id="root"
>
  <div>
    <div
      class="a"
    >
      <div
        class="b"
      />
    </div>
  </div>
</div>
`;

exports[`selectorToElement should append "div .a:last-child .b" to the root element 1`] = `
<div
  id="root"
>
  <div>
    <div
      class="a"
    >
      <div
        class="b"
      />
    </div>
  </div>
</div>
`;

exports[`selectorToElement should append "div .a:nth-child(6)" to the root element 1`] = `
<div
  id="root"
>
  <div>
    <div />
    <div />
    <div />
    <div />
    <div />
    <div
      class="a"
    />
  </div>
</div>
`;

exports[`selectorToElement should append "div .a[b=c] ul.list" to the root element 1`] = `
<div
  id="root"
>
  <div>
    <div
      b="c"
      class="a"
    >
      <ul
        class="list"
      />
    </div>
  </div>
</div>
`;

exports[`selectorToElement should append "div [a=b] #c:nth-child(5n + 2)" to the root element 1`] = `
<div
  id="root"
>
  <div>
    <div
      a="b"
    >
      <div />
      <div
        id="c"
      />
    </div>
  </div>
</div>
`;

exports[`selectorToElement should append "div ~ span" to the root element 1`] = `
<div
  id="root"
>
  <div />
  <span />
</div>
`;

exports[`selectorToElement should append "div[a~="sdad"] div[b*=lsd] div[c|=fd][d^=sdu]" to the root element 1`] = `
<div
  id="root"
>
  <div
    a="sdad"
  >
    <div
      b="lsd"
    >
      <div
        c="fd"
        d="sdu"
      />
    </div>
  </div>
</div>
`;

exports[`selectorToElement should append "fieldset input:invalid" to the root element 1`] = `
<div
  id="root"
>
  <fieldset>
    <input />
  </fieldset>
</div>
`;

exports[`selectorToElement should append "fieldset input:optional" to the root element 1`] = `
<div
  id="root"
>
  <fieldset>
    <input />
  </fieldset>
</div>
`;

exports[`selectorToElement should append "fieldset input:required" to the root element 1`] = `
<div
  id="root"
>
  <fieldset>
    <input
      required=""
    />
  </fieldset>
</div>
`;

exports[`selectorToElement should append "fieldset option:checked" to the root element 1`] = `
<div
  id="root"
>
  <fieldset>
    <option />
  </fieldset>
</div>
`;

exports[`selectorToElement should append "fieldset select:disabled" to the root element 1`] = `
<div
  id="root"
>
  <fieldset>
    <select
      disabled=""
    />
  </fieldset>
</div>
`;

exports[`selectorToElement should append "fieldset textarea:valid" to the root element 1`] = `
<div
  id="root"
>
  <fieldset>
    <textarea />
  </fieldset>
</div>
`;

exports[`selectorToElement should append "ol.a li.b.c#test div a article#art1.test[data-test="testing"]" to the root element 1`] = `
<div
  id="root"
>
  <ol
    class="a"
  >
    <li
      class="b c"
      id="test"
    >
      <div>
        <a>
          <article
            class="test"
            data-test="testing"
            id="art1"
          />
        </a>
      </div>
    </li>
  </ol>
</div>
`;

exports[`selectorToElement should append branch to a specified root 1`] = `
<span>
  <div />
  <span>
    <div
      b="c"
      class="a"
    >
      <ul
        class="list"
      />
    </div>
  </span>
</span>
`;

exports[`selectorToElement should support the :has pseudo class 1`] = `
<article>
  <div />
  <span>
    <div
      b="c"
      class="a"
    >
      <div
        class="test"
        id="oi"
      >
        <div>
          <ul
            class="list"
          >
            <li
              data-item="testing"
            />
          </ul>
        </div>
        <button
          class="sa fsd"
        />
      </div>
    </div>
  </span>
</article>
`;
